Add support for MPOpenAPI custom endpoints for proxies

abutch3r opened the pull request:
Retrieve Config to get the MPOpenAPI UI and Document endpoints and use those for comparing against a `referer` header and improve checks if custom endpoints used. if default values for the endpoints then old behavior is maintained

Add proxy test for customized endpoints - existing Proxy tests cover defaults and wider values

Fix OpenAPIEndpointManager imports

fixes #25439

abutch3r committed:
Add support for MPOpenAPI custom endpoints for proxies

Retrieve Config to get the MPOpenAPI UI and Document endpoints and use those for referer headers and improve checks if custom endpoints used. if default old behavior is maintained

Add proxy test for customized endpoints - existing Proxy tests cover defaults and wider values

Fix OpenAPIEndpointManager imports

abutch3r commented:
#build

Azquelt commented on dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/utils/ProxySupportUtil.java:
```diff
@@ -33,13 +36,45 @@ public class ProxySupportUtil {
         String urlString;
         String refererHeader = request.getHeader(HTTP_HEADER_REFERER);
+
+        String docPath = "/openapi";
+        String uiPath = docPath+"/ui";
```
- We should have constants somewhere for the default paths
- Code has not been formatted.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/utils/ProxySupportUtil.java:
```diff
@@ -33,13 +36,45 @@ public class ProxySupportUtil {
+        // retrieve the context from the request so that we can retrieve the values of the OpenAPI endpoints
+        BundleContext bundleContext = (BundleContext) request.getServletContext().getAttribute("osgi-bundlecontext");
+        //ensure we have a context in case for some reason this is activated with the bundle having stopped
+        if(bundleContext != null) {
+            ServiceTracker<OpenAPIEndpointProvider,OpenAPIEndpointProvider> openAPIEndpointTracker = new ServiceTracker<>(bundleContext, OpenAPIEndpointProvider.class, null);
+            openAPIEndpointTracker.open();
+            docPath = openAPIEndpointTracker.getService().getOpenAPIDocUrl();
+            uiPath = openAPIEndpointTracker.getService().getOpenAPIUIUrl();
+            openAPIEndpointTracker.close();
+        }
```
There's no point having a `ServiceTracker` if you're going to open and immediately close it. The idea of the `ServiceTracker` is that it's long-lived and watches for changes so that we avoid having to look up the service every time.

If you don't have a long-lived object to attach the `ServiceTracker` to, just look up the service directly with [BundleContext.getService](https://docs.osgi.org/javadoc/osgi.core/7.0.0/org/osgi/framework/BundleContext.html#getService-org.osgi.framework.ServiceReference-).

Whether you use a tracker or `BundleContext.getService`, you still need to check whether the returned service is `null`.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/fat/config/OpenAPIConfigTest.java:
```diff
@@ -255,6 +258,53 @@ public void testApplicationPathConfict() throws Exception {
+        assertTrue("Check that the list of servers has single entries",model.getServers().size() == 1);
+        assertTrue("Check host reverts to the requestUrl host", model.getServers().get(0).getUrl().contains("http://" + server.getHostname()+ ":" + server.getHttpDefaultPort() + "/" + APP_NAME));
+
+        //path does not match, but ends in `/ui`, so should revert to default host, however as port differs from server ports, only a single entry should be being returned
+        model = new OpenAPIConnection(server, "/foo").header("Referer","http://testurl4/random/ui").downloadModel();
+        // Path mismatch with `/ui`, should revert to server host and server http port - https not enabled so only single server should be returned
+        assertTrue("Check that the list of servers has a single entries",model.getServers().size() == 1);
+        assertTrue("Check host reverts to the requestUrl host", model.getServers().get(0).getUrl().contains("http://"+server.getHostname() + ":" + server.getHttpDefaultPort() + "/" + APP_NAME));
```
The two comments here are saying that the expected result occurs for different reasons - which is correct?

abutch3r replied:
Addressed comments left over from previously larger test case to clarify what we are doing in the tests and the reason for some of our expectations for number of returned servers

Azquelt commented on dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/fat/config/OpenAPIConfigTest.java:
```diff
@@ -255,6 +258,53 @@ public void testApplicationPathConfict() throws Exception {
+        //Default HTTP port with matching docPath
+        OpenAPI model = new OpenAPIConnection(server, "/foo").header("Referer","http://testurl1/foo").downloadModel();
+        //as ports is not a liberty report, only return a single server
+        assertTrue("Check that the list of servers is not empty",model.getServers().size()==1);
+        //check that the server name has changed to allow for either http:// or https:// to be first
+        assertTrue("check that the servers entry use the server from the referer header",model.getServers().get(0).getUrl().contains("http://testurl1/"+APP_NAME));
```
Use `assertThat` rather than `assertTrue` so that you get better error messages.

abutch3r replied:
All `assertTrue` statements have been replaced with `assertThat`

abutch3r committed:
Replace use of ServiceTracker with recommended service retrieval

Add in default values to proxy classes to make it easier to compared to defaults.

Address test comment comments and replace assertTrue with assertThat and address formatting issues

abutch3r commented:
#libby #build

Azquelt commented on dev/com.ibm.ws.microprofile.openapi/src/com/ibm/ws/microprofile/openapi/utils/ProxySupportUtil.java:
```diff
@@ -12,40 +12,87 @@
                 urlString = request.getRequestURL().toString();
+                if (OpenAPIUtils.isEventEnabled(tc)) {
+                    Tr.warning(tc, "Unable to use Referer header, using request url to generate servers: " + urlString);
+                }
```
Should be `Tr.event` rather than `Tr.warning`

Azquelt replied:
It's probably useful to output a trace when we use the referrer header as well.

Azquelt commented on dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/fat/config/OpenAPIConfigTest.java:
```diff
@@ -249,12 +257,76 @@ public void testApplicationPathConfict() throws Exception {
+        // Default HTTP protocol port with different host and matching docPath returns
+        // single server entry
+        OpenAPI model = new OpenAPIConnection(server, "/foo").header("Referer", "http://testurl1/foo").downloadModel();
+        assertThat(model.getServers().size(), Matchers.is(1));
```
```suggestion
        assertThat(model.getServers(), Matchers.hasSize(1));
```

Azquelt commented on dev/com.ibm.ws.microprofile.openapi_fat/fat/src/com/ibm/ws/microprofile/openapi/fat/config/OpenAPIConfigTest.java:
```diff
@@ -249,12 +257,76 @@ public void testApplicationPathConfict() throws Exception {
+
+        // Default HTTPS protocol port with matching uiPath returns single server entry
+        model = new OpenAPIConnection(server, "/foo").header("Referer", "https://testurl2/bar").downloadModel();
+        assertThat(model.getServers().size(), Matchers.is(1));
```
```suggestion
        assertThat(model.getServers(), Matchers.hasSize(1));
```

Azquelt commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/utils/ProxySupportUtil.java:
```diff
@@ -50,19 +53,62 @@ public static void processRequest(HttpServletRequest request, ServerInfo serverI
                 urlString = request.getRequestURL().toString();
+                if (LoggingUtils.isEventEnabled(tc)) {
+                    Tr.warning(tc, "Unable to use Referer header, using request url to generate servers: " + urlString);
+                }
```
```suggestion
                if (LoggingUtils.isEventEnabled(tc)) {
                    Tr.event(tc, "Unable to use Referer header, using request url to generate servers: " + urlString);
                }
```

Azquelt replied:
Can also add trace for the case where we use the refererHeader here.

Azquelt reviewed (changes requested):
This is looking really good. Only serious problem is use of `Tr.warning` where it should be `Tr.event`.

abutch3r committed:
replace log warnings with events

Add event to confirm use of referer header given we log use of request url

Update test to use hasSize inplace of is

abutch3r commented:
#libby

abutch3r commented:
#build

abutch3r commented:
#build

abutch3r commented:
#build
